ui/hooks/useNftCollectionsMetadata.ts

```diff
@@ -0,0 +1,64 @@
+import { useEffect, useMemo, useState } from 'react';
+import { NameType } from '@metamask/name-controller';
+import { Collection } from '@metamask/assets-controllers';
+import { useSelector } from 'react-redux';
+import { getCurrentChainId } from '../selectors';
+import { fetchNftCollectionsMetadata } from '../store/actions';
+
+export type UseNftCollectionsMetadataRequest = {
+  value: string;
+  type: NameType;
+};
+
+type CollectionsData = {
+  [key: string]: Collection;
+};
+
+export function useNftCollectionsMetadata(
+  requests: UseNftCollectionsMetadataRequest[],
+) {
+  const chainId = useSelector(getCurrentChainId);
+  const [collectionsMetadata, setCollectionsMetadata] =
+    useState<CollectionsData>({});
+
+  const memoisedRequestsWithChainId = useMemo(() => {
+    return {
+      contracts: requests
+        .filter(({ value }) => value)
+        .map(({ value }) => value),
+      chainId,
+    };
+  }, [JSON.stringify(requests), chainId]);
+
+  useEffect(() => {
+    const fetchCollections = async () => {
+      try {
+        const collectionsMetadata = await fetchNftCollectionsMetadata(
+          memoisedRequestsWithChainId.contracts,
+          memoisedRequestsWithChainId.chainId,
+        );
+
+        const collectionsData: CollectionsData =
+          collectionsMetadata.collections.reduce(
+            (acc: CollectionsData, collection) => {
+              // This cast is necessary because the id is set as optional string in the Collection type
+              const collectionKey = (collection.id as string).toLowerCase();
+              acc[collectionKey] = collection;
+              return acc;
+            },
+            {},
+          );
+
+        setCollectionsMetadata(collectionsData);
```

Comment:
If load becomes an issue, we could avoid some duplicate calls by caching the results in the client in a Redux slice?

Alternatively, assuming the limits are on the Reservoir side, could we create a custom endpoint specifically for the spam check that caches all the results from Reservoir?

@bergeron

Reply:
If we choose caching, could this caching happen in the controller to prevent code duplication in the clients?